File: modules/vlc-player/ios/VlcPlayerView.swift
import ExpoModulesCore
import MobileVLCKit
import UIKit

class VlcPlayerView: ExpoView {
    private var mediaPlayer: VLCMediaPlayer?
    private var videoView: UIView?
    private var progressUpdateInterval: TimeInterval = 1.0  // Update interval set to 1 second
    private var isPaused: Bool = false
    private var currentGeometryCString: [CChar]?
    private var lastReportedState: VLCMediaPlayerState?
    private var lastReportedIsPlaying: Bool?
    private var customSubtitles: [(internalName: String, originalName: String)] = []
    private var startPosition: Int32 = 0
    private var isMediaReady: Bool = false
    private var externalTrack: [String: String]?
    private var progressTimer: DispatchSourceTimer?
    private var isStopping: Bool = false  // Define isStopping here
    var hasSource = false

    // MARK: - Initialization

    required init(appContext: AppContext? = nil) {
        super.init(appContext: appContext)
        setupView()
        setupNotifications()
        setupProgressTimer()
    }

    // MARK: - Setup

    private func setupView() {
        DispatchQueue.main.async {
            self.backgroundColor = .black
            self.videoView = UIView()
            self.videoView?.translatesAutoresizingMaskIntoConstraints = false

            if let videoView = self.videoView {
                self.addSubview(videoView)
                NSLayoutConstraint.activate([
                    videoView.leadingAnchor.constraint(equalTo: self.leadingAnchor),
                    videoView.trailingAnchor.constraint(equalTo: self.trailingAnchor),
                    videoView.topAnchor.constraint(equalTo: self.topAnchor),
                    videoView.bottomAnchor.constraint(equalTo: self.bottomAnchor),
                ])
            }
        }
    }

    private func setupNotifications() {
        NotificationCenter.default.addObserver(
            self, selector: #selector(applicationWillResignActive),
            name: UIApplication.willResignActiveNotification, object: nil)
        NotificationCenter.default.addObserver(
            self, selector: #selector(applicationDidBecomeActive),
            name: UIApplication.didBecomeActiveNotification, object: nil)
    }

    private func setupProgressTimer() {
        progressTimer = DispatchSource.makeTimerSource(queue: DispatchQueue.main)
        progressTimer?.schedule(deadline: .now(), repeating: progressUpdateInterval)
        progressTimer?.setEventHandler { [weak self] in
            self?.updateVideoProgress()
        }
        progressTimer?.resume()
    }

    // MARK: - Public Methods

    @objc func play() {
        DispatchQueue.main.async { [weak self] in
            self?.mediaPlayer?.play()
            self?.isPaused = false
            print("Play")
        }
    }

    @objc func pause() {
        DispatchQueue.main.async { [weak self] in
            self?.mediaPlayer?.pause()
            self?.isPaused = true
        }
    }

    @objc func seekTo(_ time: Int32) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self, let player = self.mediaPlayer else { return }

            let wasPlaying = player.isPlaying
            if wasPlaying {
                player.pause()
            }

            if let duration = player.media?.length.intValue {
                print("Seeking to time: \(time) Video Duration \(duration)")

                // If the specified time is greater than the duration, seek to the end
                let seekTime = time > duration ? duration - 1000 : time
                player.time = VLCTime(int: seekTime)

                // Wait for a short moment to ensure the seek has been processed
                DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
                    if wasPlaying {
                        player.play()
                    }
                    self.updatePlayerState()
                }
            } else {
                print("Error: Unable to retrieve video duration")
            }
        }
    }

    @objc func setSource(_ source: [String: Any]) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }

            let mediaOptions = source["mediaOptions"] as? [String: Any] ?? [:]
            self.externalTrack = source["externalTrack"] as? [String: String]
            var initOptions = source["initOptions"] as? [Any] ?? []
            self.startPosition = source["startPosition"] as? Int32 ?? 0
            initOptions.append("--start-time=\(self.startPosition)")

            guard let uri = source["uri"] as? String, !uri.isEmpty else {
                print("Error: Invalid or empty URI")
                self.onVideoError?(["error": "Invalid or empty URI"])
                return
            }

            let autoplay = source["autoplay"] as? Bool ?? false
            let isNetwork = source["isNetwork"] as? Bool ?? false

            self.onVideoLoadStart?(["target": self.reactTag ?? NSNull()])
            self.mediaPlayer = VLCMediaPlayer(options: initOptions)
            self.mediaPlayer?.delegate = self
            self.mediaPlayer?.drawable = self.videoView
            self.mediaPlayer?.scaleFactor = 0

            let media: VLCMedia
            if isNetwork {
                print("Loading network file: \(uri)")
                media = VLCMedia(url: URL(string: uri)!)
            } else {
                print("Loading local file: \(uri)")
                if uri.starts(with: "file://"), let url = URL(string: uri) {
                    media = VLCMedia(url: url)
                } else {
                    media = VLCMedia(path: uri)
                }
            }

            print("Debug: Media options: \(mediaOptions)")
            media.addOptions(mediaOptions)

            // Apply subtitle options
            let subtitleTrackIndex = source["subtitleTrackIndex"] as? Int ?? -1
            print("Debug: Subtitle track index from source: \(subtitleTrackIndex)")
            self.setSubtitleTrack(subtitleTrackIndex)

            self.mediaPlayer?.media = media
            hasSource = true

            if autoplay {
                print("Playing...")
                self.play()
            }
        }
    }

    @objc func setAudioTrack(_ trackIndex: Int) {
        DispatchQueue.main.async {
            self.mediaPlayer?.currentAudioTrackIndex = Int32(trackIndex)
        }
    }

    @objc func getAudioTracks() -> [[String: Any]]? {
        guard let trackNames = mediaPlayer?.audioTrackNames,
            let trackIndexes = mediaPlayer?.audioTrackIndexes
        else {
            return nil
        }

        return zip(trackNames, trackIndexes).map { name, index in
            return ["name": name, "index": index]
        }
    }

    @objc func setSubtitleTrack(_ trackIndex: Int) {
        print("Debug: Attempting to set subtitle track to index: \(trackIndex)")
        DispatchQueue.main.async {
            self.mediaPlayer?.currentVideoSubTitleIndex = Int32(trackIndex)
            print(
                "Debug: Current subtitle track index after setting: \(self.mediaPlayer?.currentVideoSubTitleIndex ?? -1)"
            )
        }
    }

    @objc func setSubtitleURL(_ subtitleURL: String, name: String) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self, let url = URL(string: subtitleURL) else {
                print("Error: Invalid subtitle URL")
                return
            }

            let result = self.mediaPlayer?.addPlaybackSlave(url, type: .subtitle, enforce: true)
            if let result = result {
                let internalName = "Track \(self.customSubtitles.count + 1)"
                print("Subtitle added with result: \(result) \(internalName)")
                self.customSubtitles.append((internalName: internalName, originalName: name))
            } else {
                print("Failed to add subtitle")
            }
        }
    }

    @objc func getSubtitleTracks() -> [[String: Any]]? {
        guard let mediaPlayer = self.mediaPlayer else {
            return nil
        }

        let count = mediaPlayer.numberOfSubtitlesTracks
        print("Debug: Number of subtitle tracks: \(count)")

        guard count > 0 else {
            return nil
        }

        var tracks: [[String: Any]] = []

        if let names = mediaPlayer.videoSubTitlesNames as? [String],
            let indexes = mediaPlayer.videoSubTitlesIndexes as? [NSNumber]
        {
            for (index, name) in zip(indexes, names) {
                if let customSubtitle = customSubtitles.first(where: { $0.internalName == name }) {
                    tracks.append(["name": customSubtitle.originalName, "index": index.intValue])
                } else {
                    tracks.append(["name": name, "index": index.intValue])
                }
            }
        }

        print("Debug: Subtitle tracks: \(tracks)")
        return tracks
    }

    private func setSubtitleTrackByName(_ trackName: String) {
        guard let mediaPlayer = self.mediaPlayer else { return }

        // Get the subtitle tracks and their indexes
        if let names = mediaPlayer.videoSubTitlesNames as? [String],
            let indexes = mediaPlayer.videoSubTitlesIndexes as? [NSNumber]
        {
            for (index, name) in zip(indexes, names) {
                if name.starts(with: trackName) {
                    let trackIndex = index.intValue
                    print("Track Index setting to: \(trackIndex)")
                    setSubtitleTrack(trackIndex)
                    return
                }
            }
        }
        print("Track not found for name: \(trackName)")
    }

    @objc func stop(completion: (() -> Void)? = nil) {
        guard !isStopping else {
            completion?()
            return
        }
        isStopping = true

        // If we're not on the main thread, dispatch to main thread
        if !Thread.isMainThread {
            DispatchQueue.main.async { [weak self] in
                self?.performStop(completion: completion)
            }
        } else {
            performStop(completion: completion)
        }
    }

    // MARK: - Private Methods

    @objc private func applicationWillResignActive() {

    }

    @objc private func applicationDidBecomeActive() {

    }

    private func performStop(completion: (() -> Void)? = nil) {
        // Stop the media player
        mediaPlayer?.stop()

        // Remove observer
        NotificationCenter.default.removeObserver(self)

        // Clear the video view
        videoView?.removeFromSuperview()
        videoView = nil

        // Release the media player
        mediaPlayer?.delegate = nil
        mediaPlayer = nil

        isStopping = false
        completion?()
    }

    private func updateVideoProgress() {
        guard let player = self.mediaPlayer else { return }

        let currentTimeMs = player.time.intValue
        let durationMs = player.media?.length.intValue ?? 0

        if currentTimeMs >= 0 && currentTimeMs < durationMs {
            if player.isPlaying && !self.isMediaReady {
                self.isMediaReady = true
                // Set external track subtitle when starting.
                if let externalTrack = self.externalTrack {
                    if let name = externalTrack["name"], !name.isEmpty {
                        let deliveryUrl = externalTrack["DeliveryUrl"] ?? ""
                        self.setSubtitleURL(deliveryUrl, name: name)
                    }
                }
            }
            self.onVideoProgress?([
                "currentTime": currentTimeMs,
                "duration": durationMs,
            ])
        }
    }

    // MARK: - Expo Events

    @objc var onPlaybackStateChanged: RCTDirectEventBlock?
    @objc var onVideoLoadStart: RCTDirectEventBlock?
    @objc var onVideoStateChange: RCTDirectEventBlock?
    @objc var onVideoProgress: RCTDirectEventBlock?
    @objc var onVideoLoadEnd: RCTDirectEventBlock?
    @objc var onVideoError: RCTDirectEventBlock?

    // MARK: - Deinitialization

    deinit {
        performStop()
        progressTimer?.cancel()
    }
}

extension VlcPlayerView: VLCMediaPlayerDelegate {
    func mediaPlayerStateChanged(_ aNotification: Notification) {
        self.updatePlayerState()
    }

    private func updatePlayerState() {
        guard let player = self.mediaPlayer else { return }
        let currentState = player.state

        var stateInfo: [String: Any] = [
            "target": self.reactTag ?? NSNull(),
            "currentTime": player.time.intValue,
            "duration": player.media?.length.intValue ?? 0,
            "error": false,
        ]

        if player.isPlaying {
            stateInfo["isPlaying"] = true
            stateInfo["isBuffering"] = false
            stateInfo["state"] = "Playing"
        } else {
            stateInfo["isPlaying"] = false
            stateInfo["state"] = "Paused"
        }

        if player.state == VLCMediaPlayerState.buffering {
            stateInfo["isBuffering"] = true
            stateInfo["state"] = "Buffering"
        } else if player.state == VLCMediaPlayerState.error {
            print("player.state ~ error")
            stateInfo["state"] = "Error"
            self.onVideoLoadEnd?(stateInfo)
        } else if player.state == VLCMediaPlayerState.opening {
            print("player.state ~ opening")
            stateInfo["state"] = "Opening"
        }

        if self.lastReportedState != currentState
            || self.lastReportedIsPlaying != player.isPlaying
        {
            self.lastReportedState = currentState
            self.lastReportedIsPlaying = player.isPlaying
            self.onVideoStateChange?(stateInfo)
        }

    }

    func mediaPlayerTimeChanged(_ aNotification: Notification) {
        // No need to call updateVideoProgress here as it's handled by the timer
    }
}

extension VlcPlayerView: VLCMediaDelegate {
    // Implement VLCMediaDelegate methods if needed
}

extension VLCMediaPlayerState {
    var description: String {
        switch self {
        case .opening: return "Opening"
        case .buffering: return "Buffering"
        case .playing: return "Playing"
        case .paused: return "Paused"
        case .stopped: return "Stopped"
        case .ended: return "Ended"
        case .error: return "Error"
        case .esAdded: return "ESAdded"
        @unknown default: return "Unknown"
        }
    }
}

File: modules/vlc-player/android/src/main/java/expo/modules/vlcplayer/VlcPlayerView.kt
package expo.modules.vlcplayer

import android.content.Context
import android.util.Log
import android.view.ViewGroup
import android.widget.FrameLayout
import androidx.lifecycle.LifecycleObserver
import android.net.Uri
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.views.ExpoView
import expo.modules.kotlin.viewevent.EventDispatcher
import org.videolan.libvlc.LibVLC
import org.videolan.libvlc.Media
import org.videolan.libvlc.interfaces.IMedia
import org.videolan.libvlc.MediaPlayer
import org.videolan.libvlc.util.VLCVideoLayout

class VlcPlayerView(context: Context, appContext: AppContext) : ExpoView(context, appContext), LifecycleObserver, MediaPlayer.EventListener {

    private var libVLC: LibVLC? = null
    private var mediaPlayer: MediaPlayer? = null
    private lateinit var videoLayout: VLCVideoLayout
    private var isPaused: Boolean = false
    private var lastReportedState: Int? = null
    private var lastReportedIsPlaying: Boolean? = null
    private var media : Media? = null

    private val onVideoProgress by EventDispatcher()
    private val onVideoStateChange by EventDispatcher()
    private val onVideoLoadEnd by EventDispatcher()

    private var startPosition: Int? = 0
    private var isTranscodedStream: Boolean = false
    private var isMediaReady: Boolean = false
    private var externalTrack: Map<String, String>? = null

    init {
        setupView()
    }

    private fun setupView() {
        Log.d("VlcPlayerView", "Setting up view")
        setBackgroundColor(android.graphics.Color.WHITE)
        videoLayout = VLCVideoLayout(context).apply {
            layoutParams = LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT)
        }
        addView(videoLayout)
        Log.d("VlcPlayerView", "View setup complete")
    }

    fun setSource(source: Map<String, Any>) {
        val mediaOptions = source["mediaOptions"] as? Map<String, Any> ?: emptyMap()
        val autoplay = source["autoplay"] as? Boolean ?: false
        val isNetwork = source["isNetwork"] as? Boolean ?: false
        externalTrack = source["externalTrack"] as? Map<String, String>
        startPosition = (source["startPosition"] as? Double)?.toInt() ?: 0

        val initOptions = source["initOptions"] as? MutableList<String> ?: mutableListOf()
        initOptions.add("--start-time=$startPosition")


        val uri = source["uri"] as? String
        if (uri != null && uri.contains("m3u8")) {
            isTranscodedStream = true
        }

        // Handle video load start event
        // onVideoLoadStart?.invoke(mapOf("target" to reactTag ?: "null"))

        libVLC = LibVLC(context, initOptions)
        mediaPlayer = MediaPlayer(libVLC)
        mediaPlayer?.attachViews(videoLayout, null, false, false)
        mediaPlayer?.setEventListener(this)

        Log.d("VlcPlayerView", "Loading network file: $uri")
        media = Media(libVLC, Uri.parse(uri))
        mediaPlayer?.media = media


        Log.d("VlcPlayerView", "Debug: Media options: $mediaOptions")
        // media.addOptions(mediaOptions)

        // Apply subtitle options
        // val subtitleTrackIndex = source["subtitleTrackIndex"] as? Int ?: -1
        // Log.d("VlcPlayerView", "Debug: Subtitle track index from source: $subtitleTrackIndex")

        // if (subtitleTrackIndex >= -1) {
        //     setSubtitleTrack(subtitleTrackIndex)
        //     Log.d("VlcPlayerView", "Debug: Set subtitle track to index: $subtitleTrackIndex")
        // } else {
        //     Log.d("VlcPlayerView", "Debug: Subtitle track index is less than -1, not setting")
        // }


        if (autoplay) {
            Log.d("VlcPlayerView", "Playing...")
            play()
        }
    }

    fun play() {
        mediaPlayer?.play()
        isPaused = false
    }

    fun pause() {
        mediaPlayer?.pause()
        isPaused = true
    }

    fun stop() {
        mediaPlayer?.stop()
    }

    fun seekTo(time: Int) {
        mediaPlayer?.let { player ->
            val wasPlaying = player.isPlaying
            if (wasPlaying) {
                player.pause()
            }

            val duration = player.length.toInt()
            val seekTime = if (time > duration) duration - 1000 else time
            player.time = seekTime.toLong()

            if (wasPlaying) {
                player.play()
            }
        }
    }

    fun setAudioTrack(trackIndex: Int) {
        mediaPlayer?.setAudioTrack(trackIndex)
    }

    fun getAudioTracks(): List<Map<String, Any>>? {

        println("getAudioTracks")
        println(mediaPlayer?.getAudioTracks())
        val trackDescriptions = mediaPlayer?.audioTracks ?: return null

        return trackDescriptions.map { trackDescription ->
            mapOf("name" to trackDescription.name, "index" to trackDescription.id)
        }
    }

    fun setSubtitleTrack(trackIndex: Int) {
        mediaPlayer?.setSpuTrack(trackIndex)
    }

    // fun getSubtitleTracks(): List<Map<String, Any>>? {
    //     return mediaPlayer?.getSpuTracks()?.map { trackDescription ->
    //         mapOf("name" to trackDescription.name, "index" to trackDescription.id)
    //     }
    // }

    fun getSubtitleTracks(): List<Map<String, Any>>? {
        val subtitleTracks = mediaPlayer?.spuTracks?.map { trackDescription ->
            mapOf("name" to trackDescription.name, "index" to trackDescription.id)
        }

        // Debug statement to print the result
        Log.d("VlcPlayerView", "Subtitle Tracks: $subtitleTracks")

        return subtitleTracks
    }

    fun setSubtitleURL(subtitleURL: String, name: String) {
        println("Setting subtitle URL: $subtitleURL, name: $name")
        mediaPlayer?.addSlave(IMedia.Slave.Type.Subtitle, Uri.parse(subtitleURL), true)
    }

    override fun onDetachedFromWindow() {
        println("onDetachedFromWindow")
        super.onDetachedFromWindow()
        mediaPlayer?.stop()

        media?.release()
        mediaPlayer?.release()
        libVLC?.release()
        mediaPlayer = null
        media = null
        libVLC = null
    }

    override fun onEvent(event: MediaPlayer.Event) {
        when (event.type) {
            MediaPlayer.Event.Playing,
            MediaPlayer.Event.Paused,
            MediaPlayer.Event.Stopped,
            MediaPlayer.Event.Buffering,
            MediaPlayer.Event.EndReached,
            MediaPlayer.Event.EncounteredError -> updatePlayerState(event)
            MediaPlayer.Event.TimeChanged -> updateVideoProgress()
        }
    }

    private fun updatePlayerState(event: MediaPlayer.Event) {
        val player = mediaPlayer ?: return
        val currentState = event.type

        val stateInfo = mutableMapOf<String, Any>(
            "target" to "null", // Replace with actual target if needed
            "currentTime" to player.time.toInt(),
            "duration" to (player.media?.duration?.toInt() ?: 0),
            "error" to false
        )

        when (currentState) {
            MediaPlayer.Event.Playing -> {
                stateInfo["isPlaying"] = true
                stateInfo["isBuffering"] = false
                stateInfo["state"] = "Playing"
            }
            MediaPlayer.Event.Paused -> {
                stateInfo["isPlaying"] = false
                stateInfo["state"] = "Paused"
            }
            MediaPlayer.Event.Buffering -> {
                stateInfo["isBuffering"] = true
                stateInfo["state"] = "Buffering"
            }
            MediaPlayer.Event.EncounteredError -> {
                Log.e("VlcPlayerView", "player.state ~ error")
                stateInfo["state"] = "Error"
                onVideoLoadEnd(stateInfo);
            }
            MediaPlayer.Event.Opening -> {
                Log.d("VlcPlayerView", "player.state ~ opening")
                stateInfo["state"] = "Opening"
            }
        }


        if (lastReportedState != currentState || lastReportedIsPlaying != player.isPlaying) {
            lastReportedState = currentState
            lastReportedIsPlaying = player.isPlaying
            onVideoStateChange(stateInfo)
        }
    }

    // Only used for HLS transcoded streams
    private fun setSubtitleTrackByName(trackName: String) {
        val track = mediaPlayer?.getSpuTracks()?.firstOrNull { it.name.startsWith(trackName) }
        val trackIndex = track?.id ?: -1
        println("Track Index setting to: $trackIndex")
        if (trackIndex != -1) {
            setSubtitleTrack(trackIndex)
        }
    }


    private fun updateVideoProgress() {
        val player = mediaPlayer ?: return

        val currentTimeMs = player.time.toInt()
        val durationMs = player.media?.duration?.toInt() ?: 0

        if (currentTimeMs >= 0 && currentTimeMs < durationMs) {
            // Handle when VLC starts at cloest earliest segment skip to the start time, for transcoded streams.
            if (player.isPlaying && !isMediaReady) {
                isMediaReady = true
                externalTrack?.let {
                    val name = it["name"]
                    val deliveryUrl = it["DeliveryUrl"] ?: ""
                    if (!name.isNullOrEmpty()) {
                        if (!isTranscodedStream) {
                            setSubtitleURL(deliveryUrl, name)
                        }
                        else {
                            setSubtitleTrackByName(name)
                        }
                    }
                }

                if (isTranscodedStream && startPosition != 0) {
                    seekTo((startPosition ?: 0) * 1000)
                }
            }
            onVideoProgress(mapOf(
                "currentTime" to currentTimeMs,
                "duration" to durationMs
            ));
        }
    }
}